libarchive/archive_write_set_format_pax.c

```diff
@@ -1032,6 +1039,14 @@ archive_write_pax_header(struct archive_write *a,
 					archive_entry_set_symlink(entry_main,
 					    "././@LongSymLink");
 			}
+			else {
+				/* Otherwise, has non-ASCII characters; update the paths to
+				 * however they got decoded above */
+				if (hardlink != NULL) 
+					archive_entry_set_hardlink(entry_main, linkpath);
+				else
+					archive_entry_set_symlink(entry_main, linkpath);
```

Comment:
It's my vague understanding that not all character sets are supersets of ASCII, in which case we'll miss this conversion. That said, from what I can see above, we seem like we'd have the same issue with `pathname` if that were the case.

Reply:
Yes, we certainly would.

(See my comments above:  It's actually a lot worse than whether or not a character set is compatible with ASCII.  POSIX allows a filename to be _any_ sequence of non-zero bytes; they don't have to represent characters at all.  In that world, literally all you can do is preserve the bytes.)

Reply:
lmk what you think the best course of action would be here. To me, it seems like the most ideal thing to do would be to move this code up before the length/non-ASCII check (i.e. always set when it exists), but leaving it here would probably account for 99% of use-cases. I could also `ifdef` this for Windows only as well.

But from what I can gather, the [documentation here](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html) seems to suggest that unless an extended header specifying `hdrcharset=BINARY` is specified, UTF-8 is assumed. Looking at the code, this can only be the case if either the caller explicitly specified `hdrcharset=BINARY` _or_ if string conversions from earlier to UTF-8 fail, in which case it falls back and assumes `BINARY`. So my expectation would be that in POSIX "preserve the bytes" land, the caller should explicitly set `hdrcharset` to `BINARY`, in which case `linkpath` would just be a `memcpy` of the original bytes.

Reply:
Among other things, pax was an attempt by the POSIX folks to start moving towards Unicode filename support.  So the rules there are a bit different than historical POSIX formats like tar or cpio:
* If `hdrcharset=BINARY` is specified, filename information in the pax extended attributes is a byte sequence.
* Otherwise, filename information in the pax extended attributes is UTF-8
* Filename information in the legacy ustar header is always binary
* Filename information in old GNU extension headers is always binary

This code shouldn't be platform-sensitive.  Remember that it can get called with synthetic information or with archive entries that were read from some other archive produced on a different system.  So it should look at what character information is available in the archive_entry structure and work from there.

Reply:
I'm okay leaving it this way for now.  We can continue to discuss to see if we can come up with something better in the future.

Reply:
> I'm okay leaving it this way for now

Just to confirm - do you mean leaving it this way as in reverting my change pending additional discussions, or do you mean leaving it this way as in keeping what I have and discussing alternatives later.

> This code shouldn't be platform-sensitive

I agree. I just wanted to make sure that at a minimum, archives created on Windows with `hdrcharset=UTF-8` would be (1) valid, and (2) able to be read back correctly on Windows. The behavior now is that an empty string gets written for `linkpath` (perhaps an alternative may be to check for an empty/null `archive_entry_symlink` etc.)